src/nvc.rs

```diff
@@ -2,6 +2,111 @@ use instruction::*;
 use interconnect::*;
 
 use std::collections::HashSet;
+use std::fmt;
+
+#[derive(Copy, Clone, Default)]
+pub struct CacheEntry {
+    pub tag: u32,
+    pub base_addr: u32,
+    pub subblock_valid: [bool; 2],
+}
+
+impl fmt::Display for CacheEntry {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "base_addr: 0x{:08x}, tag: 0x{:08x}, sb1 valid: {:5}, sb2 valid: {:5}", self.base_addr, self.tag, self.subblock_valid[0], self.subblock_valid[1])
+    }
+}
+
+pub enum CacheResult {
+    Hit,
+    Miss,
+    Disabled,
+}
+
+pub struct Cache {
+    hits: usize,
+    misses: usize,
+    enable: bool,
+    entries: Box<[CacheEntry; 128]>,
+}
+
+impl Default for Cache {
+    fn default() -> Cache {
+        Cache {
+            hits: 0,
+            misses: 0,
+            enable: false,
+            entries: Box::new([CacheEntry::default(); 128]),
+        }
+    }
+}
+
+impl Cache {
+    pub fn clear_entries(&mut self, start: usize, mut count: usize) {
+        if start >= 128 {
+            return;
+        }
+
+        if count > 128 {
+            count = 128;
+        }
+
+        if start + count > 128 {
+            count = 128 - start;
+        }
+
+        for offset in 0..count {
+            self.entries[start + offset].tag = 0;
+            self.entries[start + offset].subblock_valid = [false; 2];
+         }
+    }
+
+    pub fn set_enable(&mut self, enable: bool) {
+        self.enable = enable;
+    }
+
+    pub fn enable(&self) -> bool {
+        return self.enable;
+    }
+
+    pub fn read_halfword(&mut self, interconnect: &mut Interconnect, addr: u32) -> (u16, CacheResult) {
+        let halfword = interconnect.read_halfword(addr);
+
+        if !self.enable {
+            return (halfword, CacheResult::Disabled);
+        }
+
+        let byte_offset = (addr & 0x07) as usize;
+        let entry = ((addr >> 3) & 0x7f) as usize;
+        let tag = addr >> 10;
+        let subblock = if byte_offset >= 4 { 1 } else { 0 };
+
+        if self.entries[entry].tag == tag {
+            if self.entries[entry].subblock_valid[subblock] {
+                self.hits += 1;
+                return (halfword, CacheResult::Hit);
+            }
+            self.entries[entry].subblock_valid[subblock] = true;
```

Comment:
aha, I see. That's why in the next section you reset all of those bits. Thanks for the clarification!